Спринт 13 add-controllers

ZnNr opened the pull request.

Commit:
Спринт 13 add-controllers

Commit:
13 спринт исправление ошибок  code cleanup
squash

Commit:
Спринт 13 add-controllers -check style

Commit:
Спринт 13 add-controllers -check style

Commit:
Спринт 13 add-controllers -check style

Commit:
13 спринт исправление ошибок  code cleanup

Commit:
13 спринт исправление ошибок  code cleanup

Commit:
13 спринт исправление ошибок  code cleanup

chehovnp commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,62 @@
+    @PostMapping
+    public ItemDto create(@RequestHeader("X-Sharer-User-Id") Long userId, @Valid @RequestBody ItemDto itemDto, Errors errors) {
+        log.info("Получен запрос POST /items");
+        if (errors.hasErrors()) {
```
ты уверен , что этот код работает ? error он от куда должен сюда прилететь , как ошибка валидации объекта ?

ZnNr replied:
если без этого
 .w.s.m.s.DefaultHandlerExceptionResolver : Resolved [org.springframework.web.bind.MissingRequestHeaderException: Required request header 'X-Sharer-User-Id' for method parameter type Long is not present]

chehovnp replied:
наверное можно написать, что required = false , т.е. не обязательный

chehovnp commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,30 @@
+
+    private Long id;
+
+    @NotBlank(message = "Название не может быть пустым!")
```
"Название не может быть пустым!" - эта же фраза есть в модели..

ZnNr replied:
убрал сообщение

chehovnp commented on src/main/java/ru/practicum/shareit/item/storage/ItemStorageImpl.java:
```diff
@@ -0,0 +1,50 @@
+    @Override
+    public Item add(Item item) {
+        ++id;
+        item.setId(id);
```
можно сразу item.setId(++id);

ZnNr replied:
исправил

chehovnp commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+            checkIsEmailUnique(user);
+            updatedUser.setEmail(user.getEmail());
+        }
+        log.info("Пользователь с id = {} успешно обновлен", user.getId());
```
пользователь на этой строке еще не обновлен,  это ниже происходит

chehovnp commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    public void delete(Long id) {
+        get(id);
+        log.info("Пользователь с id = {} успешно удален", id);
```
еще не удален , может и не удалиться

ZnNr replied:
поправил

chehovnp commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private void checkIsEmailUnique(User user) {
+        for (User u : userStorage.findAll()) {
```
пока видится проверка не достаточно оптимальной.  
можно добавить для хранения всех емайлов SET

chehovnp commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Override
+    public void delete(Long id) {
+        get(id);
```
а что делает этот вызов get(id);?

ZnNr replied:
звал айди для логгера...

chehovnp replied:
все равно смысла не вижу (
хочешь логировать логируй,  чего звать то )

chehovnp commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,67 @@
+
+    @PatchMapping("/{id}")
+    public UserDto update(@RequestBody User newUser, @PathVariable long id, Errors errors) {
+        log.info("Получен запрос PATCH /users/id");
```
почему ниже в методе сам id выводится , а  в этом нет 
"Получен запрос DELETE /users/{}", id 

ты же один пишешь этот код,  выбери стиль и соответствуй ему.

ZnNr replied:
исправил на         log.info("Получен запрос DELETE /users/"+ id);

chehovnp reviewed (commented):
Zinnur, привет.
оставил комментарии на что обратить внимание. 

Еще вопрос про ведение git. 
задача одна , при этом вижу 7 коммитов и все имеют одинаковые названия . 
почитай про ammend commit, force push и squash, rebase . важные инструменты которые делают разработку понятнее и опрятнее.

Commit:
Sprint 13 add-controllers Second Review - correction after review

Commit:
 This is a combination of 2 commits.
 This is the 1st commit message:

"Sprint 13 add-controllers Second Review - correction after review"

 The commit message #2 will be skipped:

 Revert " Sprint 13 add-controllers Second Review - correction after review"

 This reverts commit 2593ecfff02660017fe1dab5354a1243826b6e15.

Commit:
Merge remote-tracking branch 'origin/add-controllers' into add-controllers

Commit:
исправлен импорт toUserDto

Commit:
исправлены замечания чек стаил в файле итем контроллер

ZinnurShamsutdinov commented:
Доброго времени суток! по поводу  squash коммитов, в git bash я еще не до конца уверен в своих навыках синтаксиса bash поэтому там боюсь что то сделать с комитами необратимое... попробовал сосквошить в идее, не уверен что получилось то что нужно, появились комиты сквоша но старые коммиты не исчезли..
По проекту в целом исправил основные замечания

chehovnp commented:
да , с коммитами что -то не то . предлагаю потренироваться на кошках ))
я делаю  squash в идее,  синтаксис bash я применяю крайне редко , с гуглом ..

но веду свою разработку в одном коммите (если ветка чисто моя, не совместная), дописывая в каждой итерации новое действие.  По итогу один коммит с 5-7 пунктов что сделано.
